runtime/POSIX/stubs.c

```diff
@@ -590,7 +596,30 @@ ssize_t readahead(int fd, off64_t *offset, size_t count) {
 
 void *mmap(void *start, size_t length, int prot, int flags, int fd, off_t offset) __attribute__((weak));
 void *mmap(void *start, size_t length, int prot, int flags, int fd, off_t offset) {
-  klee_warning("ignoring (EPERM)");
+  exe_file_t* f = __get_file(fd);
+  if (f && __exe_fs.sym_pmem && f->dfile == __exe_fs.sym_pmem) {
+	  exe_disk_file_t* df = f->dfile;
+	  if (!df || !df->contents || !df->size) {
+		  klee_warning("pmem file not opened prior to mapping");
+	  } else if (offset % 4096 != 0) {
+		  klee_warning("mmap invoked without a page-aligned offset");
+	  } else {
+		  // round up length to page multiple
+		  size_t actual_length = (length % 4096 == 0 ? length : length + 4096 - length % 4096);
+		  if (offset + actual_length > df->size) {
+			  klee_warning("trying to map beyond the file size!");
+		  } else {
+#if 0
+			  printf("mmapping the pmem file at offset %lld, with length %lld (rounded from %lld).\n", 
+					  offset, actual_length, length);
+			  printf("mmap at address %p\n", (void*) (df->contents + offset));
+#endif
+			  return (void*)  (df->contents + offset);
+		  }
+	  }
+  } else {
+	  klee_warning("ignoring (EPERM), since non file-backed or not a pmem file");
+  }
```

Comment:
Nitpick, but let's be consistent with KLEE's existing spacing style if possible.

Reply:
agreed